### PR TITLE
Update to aragonOS 4

### PR DESCRIPTION
**Pull request by sohkai**
Completed:

- [x] Restructuring aragonOS section to contain the new and old reference documentation
- [x] aragonOS 4 migration guide
- [x] Small edits to APM introduction
- [x] Move APM reference documentation to APM section
- [x] Update reference docs for aragonOS 4
- [x] Review #48 
- [x] Finish `AragonApp` section
- [x] Run solidity-docgen and add some links to their API documents (e.g. `Kernel`, `ACL`, etc.)

### CODE REVIEW

**Comment by cjyabraham** [comment]
Looks good.  Once this all gets merged into merge-derek-branch2 and then to master, I can do a pass at it to check the language etc. as I have been doing with the rest of hack.aragon.org.

**Comment by sohkai** [comment]
@cjyabraham This is 99% ready to be looked at, I just need to flesh out one section in the reference doc and will push that soon.

File: docs/os-intro.md
Comment (cjyabraham) [inline docs/os-intro.md]: I think we have too many styles in this section.  I think at least the italics should be removed.

File: docs/os-motivation.md
Comment (cjyabraham) [inline docs/os-motivation.md]: could we provide a link to The Dao so it's clear what's being talked about?

**Review by cjyabraham — CHANGES_REQUESTED** [review]
A few comments.  I have not yet reviewed os-3-ref.md and os-ref.md yet but everything else has been reviewed.  I also made a commit (see below) to tighten the language and fix some small things.

File: docs/os-4-migration.md
Comment (sohkai) [reply docs/os-4-migration.md]: Has not been, but planning to update it once we get the link URL.

File: docs/apm-ref.md
Comment (sohkai) [reply docs/apm-ref.md]: Ahh, I moved this over from the os-ref document. I'll fix the links in this doc.

File: docs/apm-ref.md
Comment (sohkai) [reply docs/apm-ref.md]: See 9cae88b

File: docs/apm-ref.md
Comment (sohkai) [reply docs/apm-ref.md]: See 9cae88b

File: docs/apm-ref.md
Comment (sohkai) [reply docs/apm-ref.md]: See 25a6682

File: docs/os-intro.md
Comment (sohkai) [reply docs/os-intro.md]: See 73dc3f3.

File: docs/os-motivation.md
Comment (sohkai) [reply docs/os-motivation.md]: See 9369fbe

File: docs/apm-ref.md
Comment (sohkai) [reply docs/apm-ref.md]: Fixed with f7b0ba0

File: docs/apm-ref.md
Comment (sohkai) [reply docs/apm-ref.md]: Fixed with f7b0ba0

**Comment by sohkai** [comment]
@cjyabraham Ready to be reviewed again :). Finished the os-ref document.

File: docs/apm-ref.md
Comment (cjyabraham) [reply docs/apm-ref.md]: Some of the text still isn't readable.  I think all font sizes could go up a few notches.  The same goes for [this](https://github.com/aragon/hack/blob/9cae88bb7590384edcb9e2e2a1a2fa90c56b487f/docs/assets/apm-repo.png) diagram.  Actually, would you like me to do an edit of all the diagrams?  I think they could all be improved.  Or do we have a designer that can take a pass at them?

File: docs/os-ref.md
Comment (cjyabraham) [inline docs/os-ref.md:327]: instead of "left-shifted 32 bits to the left each" how about "each left-shifted 32 bits"?

File: docs/os-ref.md
Comment (cjyabraham) [inline docs/os-ref.md:441]: could we increase the font sizes on this diagram too?

Comment (cjyabraham) [reply docs/os-ref.md:441]: maybe I could work on all the diagrams as a batch?  Do you have the source files for all of them and what application do you use to edit?

File: docs/os-ref.md
Comment (cjyabraham) [inline docs/os-ref.md]: Is this a heading?  I'm confused by it... Is it needed?

File: docs/os-ref.md
Comment (cjyabraham) [inline docs/os-ref.md]: which source code file is this referring to?  can we link to it?

**Review by cjyabraham — CHANGES_REQUESTED** [review]

**Comment by cjyabraham** [comment]
See changeset above for some language changes.  Only os-3-ref.md remains for me to review.

**Review by luisivan — APPROVED** [review]
Looks great to me, I think it should be fine when the stuff @cjyabraham pointed out is solved

File: docs/os-ref.md
Comment (sohkai) [reply docs/os-ref.md:441]: I'm not sure we still have these somewhere, especially since this one was animated... @izqui made them originally and might.

File: docs/os-ref.md
Comment (sohkai) [inline docs/os-ref.md]: @cjyabraham I wouldn't necessarily group unstructured storage with the `DelegateProxy`, as unstructured storage is technically only one possible implementation of the `DelegateProxy` pattern.

File: docs/os-ref.md
Comment (sohkai) [inline docs/os-ref.md]: I would reword this a little bit, due to the duplicate "this"

File: docs/apm-ref.md
Comment (sohkai) [reply docs/apm-ref.md]: cc @luisivan could we have Adri take a pass through them?

I don't think we need to wait for this to merge, as it could definitely be done later when Adri has some time.

File: docs/apm-ref.md
Comment (sohkai) [reply docs/apm-ref.md]: Mmm yeah, I think we're still not being clear enough then haha.

In more words, the point this bit is trying to say is that each major version is 1-1 related to a contract address. If you query by contract address, you'll get the latest release for that major version (which may not be the latest major version), making it easy to get the latest frontend for an app if you know its contract address.

File: docs/os-ref.md
Comment (sohkai) [reply docs/os-ref.md]: Fixed in 63c5542

File: docs/os-ref.md
Comment (sohkai) [reply docs/os-ref.md]: Fixed in 63c5542

File: docs/os-ref.md
Comment (sohkai) [reply docs/os-ref.md:300]: Fixed in 63c5542

File: docs/os-ref.md
Comment (sohkai) [reply docs/os-ref.md:327]: Fixed in 63c5542

File: docs/os-ref.md
Comment (sohkai) [reply docs/os-ref.md]: Fixed in 63c5542 (wasn't needed, good catch!)

File: docs/os-ref.md
Comment (sohkai) [reply docs/os-ref.md]: Fixed in 63c5542

File: docs/apm-ref.md
Comment (cjyabraham) [reply docs/apm-ref.md]: I've created a [new issue](https://github.com/aragon/hack/issues/57) for this.

File: docs/apm-ref.md
Comment (cjyabraham) [reply docs/apm-ref.md]: I think my suggestion above was unclear.  What I'm suggestion is this (my addition is bolded below):

Fetching the latest version by contract address is useful for clients to easily fetch the latest minor or patch release associated with a particular major version. This makes it particularly easy for organizations to still be served the relevant frontends if they have not updated their smart contract code to the latest **major** version.

File: docs/os-ref.md
Comment (cjyabraham) [reply docs/os-ref.md]: Fixed in 4ad2ab5

File: docs/os-ref.md
Comment (cjyabraham) [reply docs/os-ref.md]: Ok, I've changed to this:
"aragonOS implements the DelegateProxy pattern with a special implementation called unstructured storage."

fixed in 4ad2ab5

**Review by cjyabraham — CHANGES_REQUESTED** [review]

**Comment by cjyabraham** [comment]
I've edited and submitted questions on os-3-ref.  f0b41ab and above.  I've also created [an issue](https://github.com/aragon/hack/issues/57) for improving the diagrams.

File: docs/apm-ref.md
Comment (sohkai) [reply docs/apm-ref.md]: Ahh, sorry, misread "final three words" 😅

Comment (sohkai) [reply docs/apm-ref.md]: Fixed in 12dc639

File: docs/os-3-ref.md
Comment (sohkai) [reply docs/os-3-ref.md]: Fixed in ba3bee6

File: docs/os-3-ref.md
Comment (sohkai) [reply docs/os-3-ref.md:36]: Ahh, yeah this is meant to explain more about how the proxy instance allows you to have a constant identifier (the proxy contract's address) for an application. I've reworded this in ba3bee6.

**Review by izqui — APPROVED** [review]
😍 Amazing push!

File: docs/apm.md
Comment (sohkai) [reply docs/apm.md:27]: Hmm, yeah, on some level it might. cc @cjyabraham should we not use this styling for image notes?

File: docs/stack.md
Comment (kanaagawa) [inline docs/stack.md:51]: **"Entity E can call Function F (Param P...) on App A only if Param P satisfies Rules R and Permission Manager PM can revoke or reassign that permission"**

In this statement is not clear:

What those Params are? What they represent?
Where are those Rules?
What is the Permission Manager? (it's also a little confusing name it PM)

Comment (sohkai) [reply docs/stack.md:51]: I've tried to make this a little more clear with 78dae21, although the "rules" are really hard to explain in that context (they get assigned when you create the permission).

File: docs/apm.md
Comment (cjyabraham) [reply docs/apm.md:27]: We currently don't have a way to style image captions and notes differently than warning boxes.  For notes and warnings, we use a title to introduce them so it's clear.  Captions usually don't have a title.  I think this might be ok for now but I have [made a note](https://docs.google.com/document/d/1B1nyb95LgBtrWY8Av0FnvnpUfvOb13wfjGZb28svSVs/edit#) that we need to stylistically treat them differently.  We can look at this next month when I work with Delfi on the redesign.

**Review by cjyabraham — APPROVED** [review]

File: docs/os-ref.md
Comment (kanaagawa) [inline docs/os-ref.md]: I think this isn't necessary here

File: docs/os-ref.md
Comment (kanaagawa) [inline docs/os-ref.md]:
```suggestion
> `CREATE_PERMISSION_ROLE` could be used in malicious and dangerous ways. This is initially assigned when the Kernel is first initialized. **Protect this permission.**
```

**Review by 0xGabi — CHANGES_REQUESTED** [review]

File: docs/os-ref.md
Comment (kanaagawa) [inline docs/os-ref.md]: I don't think this link is relevant here

File: docs/os-ref.md
Comment (sohkai) [reply docs/os-ref.md]: 😅 oops, forgot to change it to read `createPermission`. Fixed in 7b743b6

File: docs/os-ref.md
Comment (sohkai) [reply docs/os-ref.md]: I think having the link is still useful, since in this context it's not really clear what EVMScripts are for. I've moved the link down to the bottom of the section and reworded it in 7b743b6.

**Pull request merged**
